src/QMCWaveFunctions/LCAO/SoaAtomicBasisSet.h

```diff
@@ -665,6 +684,166 @@ struct SoaAtomicBasisSet
+
+
+    {
+      ScopedTimer local_timer(phase_timer_);
+#if not defined(QMC_COMPLEX)
+
+      PRAGMA_OFFLOAD("omp target teams distribute parallel for map(to:correctphase_ptr[:nElec]) ")
+      for (size_t i_e = 0; i_e < nElec; i_e++)
+        correctphase_ptr[i_e] = 1.0;
+
+#else
+      auto* SuperTwist_ptr = SuperTwist.data();
+
+      PRAGMA_OFFLOAD("omp target teams distribute parallel for map(to:SuperTwist_ptr[:9], \
+		     Tv_list_ptr[3*nElec*center_idx:3*nElec], correctphase_ptr[:nElec]) ")
+      for (size_t i_e = 0; i_e < nElec; i_e++)
+      {
+        //RealType phasearg = dot(3, SuperTwist.data(), 1, Tv_list.data() + 3 * i_e, 1);
+        RealType phasearg = 0;
+        for (size_t i_dim = 0; i_dim < 3; i_dim++)
+          phasearg += SuperTwist[i_dim] * Tv_list_ptr[i_dim + 3 * (i_e + center_idx * nElec)];
+        RealType s, c;
+        qmcplusplus::sincos(-phasearg, &s, &c);
+        correctphase_ptr[i_e] = ValueType(c, s);
+      }
+#endif
+    }
+
+    auto* periodic_image_displacements_ptr = periodic_image_displacements.data();
+    {
+      ScopedTimer local_timer(nelec_pbc_timer_);
+      PRAGMA_OFFLOAD("omp target teams distribute parallel for collapse(2) \
+                      map(always, to:periodic_image_displacements_ptr[:3*Nxyz]) \
+                      map(to: dr_ptr[:3*nElec*Nxyz], r_ptr[:nElec*Nxyz], displ_list_ptr[3*nElec*center_idx:3*nElec]) ")
+      for (size_t i_e = 0; i_e < nElec; i_e++)
+      {
+        for (int i_xyz = 0; i_xyz < Nxyz; i_xyz++)
+        {
+          RealType tmp_r2 = 0.0;
+          for (size_t i_dim = 0; i_dim < 3; i_dim++)
+          {
+            dr_ptr[i_dim + 3 * (i_xyz + Nxyz * i_e)] = -(displ_list_ptr[i_dim + 3 * (i_e + center_idx * nElec)] +
+                                                         periodic_image_displacements_ptr[i_dim + 3 * i_xyz]);
+            tmp_r2 += dr_ptr[i_dim + 3 * (i_xyz + Nxyz * i_e)] * dr_ptr[i_dim + 3 * (i_xyz + Nxyz * i_e)];
+          }
+          r_ptr[i_xyz + Nxyz * i_e] = std::sqrt(tmp_r2);
+        }
+      }
+    }
+
+
+    {
+      ScopedTimer local(rnl_timer_);
+      MultiRnl.batched_evaluate(r, rnl_v, Rmax);
+    }
+    // dr_new is [3 * Nxyz * nElec] realtype
+    {
+      ScopedTimer local(ylm_timer_);
+      //Ylm.mw_evaluateV(dr.data(), ylm_v.data(), Nxyz * nElec, nYlm);
+      Ylm.batched_evaluateV(dr, ylm_v);
+    }
+    ///Phase for PBC containing the phase for the nearest image displacement and the correction due to the Distance table.
+    auto* phase_fac_ptr = periodic_image_phase_factors.data();
+    auto* LM_ptr        = LM.data();
+    auto* NL_ptr        = NL.data();
+    auto* psi_ptr       = psi.data();
+
+    auto* ylm_ptr = ylm_v.data();
+    auto* rnl_ptr = rnl_v.data();
+    {
+      ScopedTimer local_timer(psi_timer_);
+      PRAGMA_OFFLOAD("omp target teams distribute parallel for collapse(2) \
+          map(always, to:phase_fac_ptr[:Nxyz], LM_ptr[:BasisSetSize], NL_ptr[:BasisSetSize]) \
```

Comment:
Could you add a FIXME saying the "always" should be removed after making the transfer only once ahead of time?

Reply:
fixed